src/webview/LeetCodeWebview.ts

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) jdneo. All rights reserved.
+// Licensed under the MIT license.
+
+import { ConfigurationChangeEvent, Disposable, ExtensionContext, ViewColumn, WebviewPanel, window, workspace } from "vscode";
+import { markdownEngine } from "./markdownEngine";
+
+export abstract class LeetCodeWebview implements Disposable {
+
+    protected panel: WebviewPanel | undefined;
+    private context: ExtensionContext;
+    private listener: Disposable;
+
+    public initialize(context: ExtensionContext): void {
+        const { onDidChangeConfiguration } = this.getWebviewOption();
+        this.context = context;
+        if (onDidChangeConfiguration) {
+            this.listener = workspace.onDidChangeConfiguration(onDidChangeConfiguration, this);
+        } else {
+            this.listener = workspace.onDidChangeConfiguration((event: ConfigurationChangeEvent) => {
+                if (event.affectsConfiguration("markdown") && this.panel) {
+                    this.panel.webview.html = this.getWebviewContent();
+                }
+            }, this);
+        }
+    }
+
+    public dispose(): void {
+        this.listener.dispose();
+        if (this.panel) {
+            this.panel.dispose();
+        }
+    }
+
+    protected showWebviewInternal(): this is { panel: WebviewPanel } {
```

Comment:
Could you explain more about this method? The return type is `this`, but the actual returned value is `true`

Reply:
`param is type` is a user-defined [type guard](https://www.typescriptlang.org/docs/handbook/advanced-types.html) in typescript's grammar. 

It reduced the scope of union type:
```ts
function isTypeDefined(obj: Type | undefined): obj is Type {
    return obj !== undefined;
}

let a: Type | undefined;
a.method() // ts error, a may be undefined

if (isTypeDefined(a)) {
    a.method(); // now a's type is only `Type`, `undefined` is reduced
}
```

`this is type` is a way to guard a class member's type:
https://github.com/Microsoft/TypeScript/issues/26212

`this is { panel: WebviewPanel }` makes sure `this.panel` is not undefined.

So the code here do not need to use `this.panel!.method()` to assert defined:
```ts
if (this.showWebviewInternal()) {
    this.panel.title = `${node.name}: Preview`; // now this.panel's undefined type is reduced
    this.panel.reveal(ViewColumn.One);
}
```

Reply:
Then why we are not just calling `showWebviewInternal()` but put it in the if block.

I do not see any benefits to do this.

Reply:
You mean
```ts
this.showWebviewInternal();
this.panel!.title = `${node.name}: Preview`; // this.panel's undefined type is not reduced
this.panel!.reveal(ViewColumn.One);
```
?

Reply:
The main problem here is, we should extract the fields in `ILeetCodeWebviewOption` as the member of the base class. The reason for doing that:

1. Each web view has its own unique `viewType`
2. Each web view has its own location to display - the `viewColumn `
3. Each web view has the member title to determine the title of the editor.

So after the extraction, what you need to do is:
1. Move all the logic about the panel update into `showWebviewInternal()`
2. Update the title field when needed.

```typescript
this.title = xxxx
this.showWebviewInternal();
```

Reply:
Ah...the next PR I prepared yesterday has done the work (panel update in `showWebviewInternal`) in another way:

In `showWebviewInternal()`:
```ts
if (!this.panel) {
    // ...old initialization code
} else {
    this.panel.title = title;
    if (viewColumn === ViewColumn.Two) {
        // Make sure second group exists. See vscode#71608 issue
        commands.executeCommand("workbench.action.focusSecondEditorGroup").then(() => {
            this.panel!.reveal(viewColumn, preserveFocus);
        });
    } else {
        this.panel.reveal(viewColumn, preserveFocus);
    }
}
```

In `show(...)`:
```ts
public async show(node: IProblem, isSideMode: boolean = false): Promise<void> {
    this.description = this.parseDescription(await leetCodeExecutor.getDescription(node), node);
    this.node = node;
    this.sideMode = isSideMode;
    this.showWebviewInternal();
    if (this.sideMode) {
        // Hide side bar for better view area
        await commands.executeCommand("workbench.action.focusSideBar").then(() => {
            commands.executeCommand("workbench.action.toggleSidebarVisibility");
        });
    }
}
```

In `getWebviewOption()`:
```ts
protected getWebviewOption(): ILeetCodeWebviewOption {
    if (!this.sideMode) { // a state when all the leetcode panels are grouped in Column 2
        return {
            viewType: "leetcode.preview",
            title: `${this.node.name}: Preview`,
            viewColumn: ViewColumn.One,
            preserveFocus: true,
        };
    } else {
        return {
            viewType: "leetcode.preview",
            title: "Description",
            viewColumn: ViewColumn.Two,
        };
    }
}
```
In this way, the data needed to update the panel is dynamically organized in `getWebviewOption()`.

Reply:
Do you mean you want to add a new setting called `isSideMode` to determine the location of the webviews?

Reply:
This is the focus of next PR, which uses column 2 to provide information when writing solution.

Reply:
So far, I cannot get the point of the method `getWebviewOption()` who else will call it?

Reply:
`getWebviewOption()` is only used in `showWebviewInternal()`, to initialized or update the panel.

Reply:
Then why just make them as the members of the class? From the code flow, the `title` is different according to the different value of `sideMode`. Actually, there is no relationship between them. The title should always update to a more meaningful value no matter where will the webview be located. That's the reason I prefer to use class member instead of `getWebviewOption()`.

Reply:
I've brought everything except `sideMode` in next PR to this PR. It includes the improvement on the webview base class, like mentioned above.

Reply:
Do you really think there is no relationship? Did you ever test yourself what is the visual effect when the title keeps `A so long long long long long name: preview` and `A so long long long long long name: solution` in Column 2? Will the user be able to easily change between the tabs?

Why do you pay more attention to something in the next PR which I haven't given illustration yet?

Reply:
I just want leetcode tabs in column 2 look like this way:
![image](https://user-images.githubusercontent.com/20227484/56185036-d512b780-604e-11e9-8dbf-534c76d0fac1.png)

Otherwise, in my small-screen PC, a `So long long long a name: Preview` just occupies all the room in the Column 2.

Reply:
Of cause I need to understand what you will do in the next PR since they have some relationship. This will determine whether we need to keep `getWebviewOption()` or make them as the class member.

The codes written in a good way should not only have good experience but also good at self-explain. This is very important to attract more community contribution. That's why I'm keeping asking you those questions. From the code itself, I cannot get your point. There's no GIF/picture to illustrate that. Then how could I know your point is that the tab is so long that make the experience bad?

Besides, From my understanding, the reason to rename to the title to `Description` is that you hope the size of the tab is not too big when VS Code has two columns. The real reason actually is having limited resolution, and multiple columns is one of the possible reason to cause the limited resolution.

Take the VS Code embedded Markdown extension as an example, You can see there is no optimization when the tab is too long.
![image](https://user-images.githubusercontent.com/6193897/56187251-9b44af80-6054-11e9-903f-13971566bb7a.png)

Reply:
I'm not saying that we should not do the optimation. It's acceptable to do it as long as the code logic is easy to understand.